File: packages/browser-domain/src/social-audit-explanation-read-model.ts
import { type Infer, Schema, withParser } from '@ocentra-parent/schema-domain/effect';
import {
  ChildProfileIdSchema,
  FamilyIdSchema,
  ParentEvidenceReferenceIdSchema,
  ParentPolicyVersionSchema,
  ParentTimestampSchema,
} from '@ocentra-parent/schema-domain/family-reference-primitives';
import {
  SocialParentPolicyActionCandidateSchema,
  SocialParentPolicyReasonCodesSchema,
} from './social-policy-compiler-values';
import {
  SocialAuditExplanationAudienceSchema,
  SocialAuditExplanationDecisionStateSchema,
  SocialAuditExplanationEventIdSchema,
  SocialAuditExplanationEvidenceKindSchema,
  type SocialAuditExplanationEvidenceKind,
  SocialAuditExplanationReadModelSchemaVersionSchema,
  SocialAuditExplanationReasonSchema,
  SocialAuditExplanationSnapshotIdSchema,
  type SocialAuditExplanationSubjectKind,
  SocialAuditExplanationSubjectKindSchema,
  SocialAuditExplanationStatusSchema,
} from './social-audit-explanation-read-model-values';

const OptionalAuditEvidenceRefSchema = Schema.Union(ParentEvidenceReferenceIdSchema, Schema.Null);
const SocialAuditExplanationRefsSchema = Schema.Array(ParentEvidenceReferenceIdSchema).pipe(
  Schema.filter((value) => value.length > 0 || 'Expected social audit/explanation refs')
);
const SocialAuditExplanationReasonsSchema = Schema.Array(SocialAuditExplanationReasonSchema).pipe(
  Schema.filter((value) => value.length > 0 || 'Expected social audit/explanation reasons')
);
const SocialAuditExplanationEvidenceLinksSchema = Schema.Array(
  Schema.Struct({
    evidenceKind: SocialAuditExplanationEvidenceKindSchema,
    evidenceRef: ParentEvidenceReferenceIdSchema,
  })
).pipe(Schema.filter((value) => value.length > 0 || 'Expected social audit/explanation evidence links'));

const SocialAuditExplanationEntryBaseSchema = Schema.Struct({
  eventId: SocialAuditExplanationEventIdSchema,
  subjectKind: SocialAuditExplanationSubjectKindSchema,
  status: SocialAuditExplanationStatusSchema,
  decisionState: SocialAuditExplanationDecisionStateSchema,
  audience: SocialAuditExplanationAudienceSchema,
  policyVersionRef: Schema.Union(ParentPolicyVersionSchema, Schema.Null),
  actionCandidate: SocialParentPolicyActionCandidateSchema,
  policyReasonCodes: SocialParentPolicyReasonCodesSchema,
  explanationReasons: SocialAuditExplanationReasonsSchema,
  evidenceLinks: SocialAuditExplanationEvidenceLinksSchema,
  auditRefs: SocialAuditExplanationRefsSchema,
  parentApprovalRequestRef: OptionalAuditEvidenceRefSchema,
  parentApprovalDecisionRef: OptionalAuditEvidenceRefSchema,
  decisionMemoryRef: OptionalAuditEvidenceRefSchema,
  connectorBoundaryRef: OptionalAuditEvidenceRefSchema,
  nativeCapabilityRef: OptionalAuditEvidenceRefSchema,
  manualRequiredRef: OptionalAuditEvidenceRefSchema,
  runtimeAuditStoreClaimed: Schema.Boolean,
  renderedExplanationUiClaimed: Schema.Boolean,
  notificationDeliveredClaimed: Schema.Boolean,
  rawAccountDataIncluded: Schema.Boolean,
  rawVideoContentIncluded: Schema.Boolean,
  rawMessageContentIncluded: Schema.Boolean,
  connectorAuthorizationClaimed: Schema.Boolean,
  nativeAppControlClaimed: Schema.Boolean,
  finalPolicyDecisionClaimed: Schema.Boolean,
  enforcementClaimed: Schema.Boolean,
});

type SocialAuditExplanationEntryCandidate = Infer<typeof SocialAuditExplanationEntryBaseSchema>;

export const SocialAuditExplanationEntrySchema = withParser(
  SocialAuditExplanationEntryBaseSchema.pipe(
    Schema.filter(
      (entry) =>
        socialAuditExplanationEntryIsHonest(entry) ||
        'Expected social audit/explanation row to stay ref-only, readable, and non-enforcing'
    )
  )
);

export const SocialAuditExplanationClaimBoundariesSchema = withParser(
  Schema.Struct({
    runtimeAuditStore: Schema.Literal('not-claimed'),
    renderedExplanationUi: Schema.Literal('not-claimed'),
    notificationDelivery: Schema.Literal('not-claimed'),
    rawAccountVideoMessageContent: Schema.Literal('not-claimed'),
    connectorAuthorization: Schema.Literal('not-claimed'),
    nativeAppControl: Schema.Literal('not-claimed'),
    finalPolicyDecision: Schema.Literal('not-claimed'),
    enforcement: Schema.Literal('not-claimed'),
  })
);

const SocialAuditExplanationSnapshotBaseSchema = Schema.Struct({
  schemaVersion: SocialAuditExplanationReadModelSchemaVersionSchema,
  snapshotId: SocialAuditExplanationSnapshotIdSchema,
  familyId: FamilyIdSchema,
  childProfileId: ChildProfileIdSchema,
  capturedAt: ParentTimestampSchema,
  entries: Schema.Array(SocialAuditExplanationEntrySchema),
  claimBoundaries: SocialAuditExplanationClaimBoundariesSchema,
});

type SocialAuditExplanationSnapshotCandidate = Infer<typeof SocialAuditExplanationSnapshotBaseSchema>;

export const SocialAuditExplanationSnapshotSchema = withParser(
  SocialAuditExplanationSnapshotBaseSchema.pipe(
    Schema.filter(
      (snapshot) =>
        socialAuditExplanationSnapshotIsHonest(snapshot) ||
        'Expected social audit/explanation snapshot to include all required readable audit rows'
    )
  )
);

export const decodeSocialAuditExplanationSnapshot = Schema.decodeUnknownSync(SocialAuditExplanationSnapshotSchema);

export type SocialAuditExplanationEntry = Infer<typeof SocialAuditExplanationEntrySchema>;
export type SocialAuditExplanationSnapshot = Infer<typeof SocialAuditExplanationSnapshotSchema>;

const RequiredSocialAuditExplanationSubjects = [
  'account-approval',
  'feed-video-gate',
  'native-app-gap',
  'connector-boundary',
  'decision-memory',
  'manual-required-gap',
] as const satisfies ReadonlyArray<SocialAuditExplanationSubjectKind>;

function socialAuditExplanationSnapshotIsHonest(snapshot: SocialAuditExplanationSnapshotCandidate): boolean {
  const subjectKinds = new Set(snapshot.entries.map((entry) => entry.subjectKind));
  return (
    subjectKinds.size === snapshot.entries.length &&
    RequiredSocialAuditExplanationSubjects.every((subjectKind) => subjectKinds.has(subjectKind))
  );
}

function socialAuditExplanationEntryIsHonest(entry: SocialAuditExplanationEntryCandidate): boolean {
  if (socialAuditExplanationEntryClaimsRuntime(entry)) {
    return false;
  }
  if (entry.status === 'ready-for-parent' && !socialAuditExplanationReadyRowIsComplete(entry)) {
    return false;
  }
  if (entry.subjectKind === 'account-approval') {
    return socialAuditExplanationAccountRowIsComplete(entry);
  }
  if (entry.subjectKind === 'feed-video-gate') {
    return socialAuditExplanationFeedVideoRowIsComplete(entry);
  }
  if (entry.subjectKind === 'decision-memory') {
    return socialAuditExplanationDecisionMemoryRowIsComplete(entry);
  }
  return socialAuditExplanationManualBoundaryRowIsComplete(entry);
}

function socialAuditExplanationReadyRowIsComplete(entry: SocialAuditExplanationEntryCandidate): boolean {
  return (
    entry.policyVersionRef !== null &&
    entry.decisionState !== 'manual-required' &&
    entry.decisionState !== 'unavailable' &&
    socialAuditExplanationHasEvidence(entry, 'policy-candidate')
  );
}

function socialAuditExplanationAccountRowIsComplete(entry: SocialAuditExplanationEntryCandidate): boolean {
  return (
    entry.parentApprovalRequestRef !== null &&
    entry.parentApprovalDecisionRef !== null &&
    socialAuditExplanationHasEvidence(entry, 'parent-approval')
  );
}

function socialAuditExplanationFeedVideoRowIsComplete(entry: SocialAuditExplanationEntryCandidate): boolean {
  return (
    entry.status === 'ready-for-parent' &&
    entry.actionCandidate !== 'unknown-candidate' &&
    socialAuditExplanationHasEvidence(entry, 'route-evidence')
  );
}

function socialAuditExplanationDecisionMemoryRowIsComplete(entry: SocialAuditExplanationEntryCandidate): boolean {
  return (
    entry.status === 'contract-only' &&
    entry.decisionState === 'candidate-only' &&
    entry.decisionMemoryRef !== null &&
    socialAuditExplanationHasEvidence(entry, 'decision-memory')
  );
}

function socialAuditExplanationManualBoundaryRowIsComplete(entry: SocialAuditExplanationEntryCandidate): boolean {
  if (entry.subjectKind === 'native-app-gap') {
    return entry.status === 'manual-required' && entry.nativeCapabilityRef !== null;
  }
  if (entry.subjectKind === 'connector-boundary') {
    return entry.status === 'manual-required' && entry.connectorBoundaryRef !== null;
  }
  return entry.status === 'manual-required' && entry.manualRequiredRef !== null;
}

function socialAuditExplanationHasEvidence(
  entry: SocialAuditExplanationEntryCandidate,
  evidenceKind: SocialAuditExplanationEvidenceKind
): boolean {
  return entry.evidenceLinks.some((evidence) => evidence.evidenceKind === evidenceKind);
}

function socialAuditExplanationEntryClaimsRuntime(entry: SocialAuditExplanationEntryCandidate): boolean {
  return (
    entry.runtimeAuditStoreClaimed ||
    entry.renderedExplanationUiClaimed ||
    entry.notificationDeliveredClaimed ||
    entry.rawAccountDataIncluded ||
    entry.rawVideoContentIncluded ||
    entry.rawMessageContentIncluded ||
    entry.connectorAuthorizationClaimed ||
    entry.nativeAppControlClaimed ||
    entry.finalPolicyDecisionClaimed ||
    entry.enforcementClaimed
  );
}

File: crates/agent-protocol/tests/contract/social_audit_explanation_read_model.rs
use ocentra_eventing::expect_value::ExpectValue;
use ocentra_parent_agent_protocol::constants;
use ocentra_parent_agent_protocol::SocialAuditExplanationClaimBoundaries;
use ocentra_parent_agent_protocol::SocialAuditExplanationEntry;
use ocentra_parent_agent_protocol::SocialAuditExplanationEvidenceLink;
use ocentra_parent_agent_protocol::SocialAuditExplanationSnapshot;
use ocentra_parent_agent_protocol::SOCIAL_AUDIT_EXPLANATION_ACTION_WARN;
use ocentra_parent_agent_protocol::SOCIAL_AUDIT_EXPLANATION_AUDIENCE_PARENT;
use ocentra_parent_agent_protocol::SOCIAL_AUDIT_EXPLANATION_CHILD_PROFILE_ID;
use ocentra_parent_agent_protocol::SOCIAL_AUDIT_EXPLANATION_CLAIM_NOT_CLAIMED;
use ocentra_parent_agent_protocol::SOCIAL_AUDIT_EXPLANATION_DECISION_CANDIDATE_ONLY;
use ocentra_parent_agent_protocol::SOCIAL_AUDIT_EXPLANATION_EVIDENCE_POLICY_CANDIDATE;
use ocentra_parent_agent_protocol::SOCIAL_AUDIT_EXPLANATION_EVIDENCE_ROUTE_EVIDENCE;
use ocentra_parent_agent_protocol::SOCIAL_AUDIT_EXPLANATION_FAMILY_ID;
use ocentra_parent_agent_protocol::SOCIAL_AUDIT_EXPLANATION_POLICY_REASON_SOCIAL_RISK_HIGH;
use ocentra_parent_agent_protocol::SOCIAL_AUDIT_EXPLANATION_POLICY_VERSION;
use ocentra_parent_agent_protocol::SOCIAL_AUDIT_EXPLANATION_REASON_EVIDENCE_LINKED;
use ocentra_parent_agent_protocol::SOCIAL_AUDIT_EXPLANATION_SCHEMA_VERSION;
use ocentra_parent_agent_protocol::SOCIAL_AUDIT_EXPLANATION_SNAPSHOT_ID;
use ocentra_parent_agent_protocol::SOCIAL_AUDIT_EXPLANATION_STATUS_READY_FOR_PARENT;
use ocentra_parent_agent_protocol::SOCIAL_AUDIT_EXPLANATION_SUBJECT_FEED_VIDEO_GATE;

#[test]
fn social_audit_explanation_snapshot_serializes_without_runtime_claims() {
    let snapshot = SocialAuditExplanationSnapshot {
        schema_version: SOCIAL_AUDIT_EXPLANATION_SCHEMA_VERSION.to_string(),
        snapshot_id: SOCIAL_AUDIT_EXPLANATION_SNAPSHOT_ID.to_string(),
        family_id: SOCIAL_AUDIT_EXPLANATION_FAMILY_ID.to_string(),
        child_profile_id: SOCIAL_AUDIT_EXPLANATION_CHILD_PROFILE_ID.to_string(),
        captured_at: constants::activity_store::TEST_TRACKING_RETENTION_DELETE_OBSERVED_AT
            .to_string(),
        entries: vec![social_audit_explanation_entry()],
        claim_boundaries: not_claimed_boundaries(),
    };

    let serialized =
        serde_json::to_value(snapshot).expect_value(constants::error::AGENT_EVENT_SERIALIZES);

    assert_eq!(
        serialized["schemaVersion"],
        SOCIAL_AUDIT_EXPLANATION_SCHEMA_VERSION
    );
    assert_eq!(
        serialized["entries"][0]["evidenceLinks"][0]["evidenceKind"],
        SOCIAL_AUDIT_EXPLANATION_EVIDENCE_ROUTE_EVIDENCE
    );
    assert_eq!(
        serialized["claimBoundaries"]["enforcement"],
        SOCIAL_AUDIT_EXPLANATION_CLAIM_NOT_CLAIMED
    );
    assert_eq!(serialized["entries"][0]["runtimeAuditStoreClaimed"], false);
}

fn not_claimed_boundaries() -> SocialAuditExplanationClaimBoundaries {
    SocialAuditExplanationClaimBoundaries {
        runtime_audit_store: SOCIAL_AUDIT_EXPLANATION_CLAIM_NOT_CLAIMED.to_string(),
        rendered_explanation_ui: SOCIAL_AUDIT_EXPLANATION_CLAIM_NOT_CLAIMED.to_string(),
        notification_delivery: SOCIAL_AUDIT_EXPLANATION_CLAIM_NOT_CLAIMED.to_string(),
        raw_account_video_message_content: SOCIAL_AUDIT_EXPLANATION_CLAIM_NOT_CLAIMED.to_string(),
        connector_authorization: SOCIAL_AUDIT_EXPLANATION_CLAIM_NOT_CLAIMED.to_string(),
        native_app_control: SOCIAL_AUDIT_EXPLANATION_CLAIM_NOT_CLAIMED.to_string(),
        final_policy_decision: SOCIAL_AUDIT_EXPLANATION_CLAIM_NOT_CLAIMED.to_string(),
        enforcement: SOCIAL_AUDIT_EXPLANATION_CLAIM_NOT_CLAIMED.to_string(),
    }
}

fn social_audit_explanation_entry() -> SocialAuditExplanationEntry {
    SocialAuditExplanationEntry {
        event_id: SOCIAL_AUDIT_EXPLANATION_SUBJECT_FEED_VIDEO_GATE.to_string(),
        subject_kind: SOCIAL_AUDIT_EXPLANATION_SUBJECT_FEED_VIDEO_GATE.to_string(),
        status: SOCIAL_AUDIT_EXPLANATION_STATUS_READY_FOR_PARENT.to_string(),
        decision_state: SOCIAL_AUDIT_EXPLANATION_DECISION_CANDIDATE_ONLY.to_string(),
        audience: SOCIAL_AUDIT_EXPLANATION_AUDIENCE_PARENT.to_string(),
        policy_version_ref: Some(SOCIAL_AUDIT_EXPLANATION_POLICY_VERSION.to_string()),
        action_candidate: SOCIAL_AUDIT_EXPLANATION_ACTION_WARN.to_string(),
        policy_reason_codes: vec![
            SOCIAL_AUDIT_EXPLANATION_POLICY_REASON_SOCIAL_RISK_HIGH.to_string()
        ],
        explanation_reasons: vec![SOCIAL_AUDIT_EXPLANATION_REASON_EVIDENCE_LINKED.to_string()],
        evidence_links: social_audit_explanation_evidence_links(),
        audit_refs: vec![
            constants::activity_store::TEST_TRACKING_EVIDENCE_REFERENCE_ID.to_string(),
        ],
        parent_approval_request_ref: None,
        parent_approval_decision_ref: None,
        decision_memory_ref: None,
        connector_boundary_ref: None,
        native_capability_ref: None,
        manual_required_ref: None,
        runtime_audit_store_claimed: false,
        rendered_explanation_ui_claimed: false,
        notification_delivered_claimed: false,
        raw_account_data_included: false,
        raw_video_content_included: false,
        raw_message_content_included: false,
        connector_authorization_claimed: false,
        native_app_control_claimed: false,
        final_policy_decision_claimed: false,
        enforcement_claimed: false,
    }
}

fn social_audit_explanation_evidence_links() -> Vec<SocialAuditExplanationEvidenceLink> {
    vec![
        SocialAuditExplanationEvidenceLink {
            evidence_kind: SOCIAL_AUDIT_EXPLANATION_EVIDENCE_ROUTE_EVIDENCE.to_string(),
            evidence_ref: format!(
                "parent-evidence-{}",
                SOCIAL_AUDIT_EXPLANATION_EVIDENCE_ROUTE_EVIDENCE
            ),
        },
        SocialAuditExplanationEvidenceLink {
            evidence_kind: SOCIAL_AUDIT_EXPLANATION_EVIDENCE_POLICY_CANDIDATE.to_string(),
            evidence_ref: format!(
                "parent-evidence-{}",
                SOCIAL_AUDIT_EXPLANATION_EVIDENCE_POLICY_CANDIDATE
            ),
        },
    ]
}
